src/useNetworkState.ts

```diff
@@ -0,0 +1,137 @@
+import { useSafeState } from './useSafeState';
+import { off, on } from './util/misc';
+
+export interface INetworkInformation extends EventTarget {
+  readonly downlink: number;
+  readonly downlinkMax: number;
+  readonly effectiveType: 'slow-2g' | '2g' | '3g' | '4g';
+  readonly rtt: number;
+  readonly saveData: boolean;
+  readonly type:
+    | 'bluetooth'
+    | 'cellular'
+    | 'ethernet'
+    | 'none'
+    | 'wifi'
+    | 'wimax'
+    | 'other'
+    | 'unknown';
+}
+
+export interface IUseNetworkState {
+  /**
+   * @desc Whether browser connected to the network or not.
+   */
+  online: boolean | undefined;
+  /**
+   * @desc Previous value of `online` property. Helps to identify if browser
+   * just connected or lost connection.
+   */
+  previous: boolean | undefined;
+  /**
+   * @desc The {Date} object pointing to the moment when state change occurred.
+   */
+  since: Date | undefined;
+  /**
+   * @desc Effective bandwidth estimate in megabits per second, rounded to the
+   * nearest multiple of 25 kilobits per seconds.
+   */
+  downlink: INetworkInformation['downlink'] | undefined;
+  /**
+   * @desc Maximum downlink speed, in megabits per second (Mbps), for the
+   * underlying connection technology
+   */
+  downlinkMax: INetworkInformation['downlinkMax'] | undefined;
+  /**
+   * @desc Effective type of the connection meaning one of 'slow-2g', '2g', '3g', or '4g'.
+   * This value is determined using a combination of recently observed round-trip time
+   * and downlink values.
+   */
+  effectiveType: INetworkInformation['effectiveType'] | undefined;
+  /**
+   * @desc Estimated effective round-trip time of the current connection, rounded
+   * to the nearest multiple of 25 milliseconds
+   */
+  rtt: INetworkInformation['rtt'] | undefined;
+  /**
+   * @desc {true} if the user has set a reduced data usage option on the user agent.
+   */
+  saveData: INetworkInformation['saveData'] | undefined;
+  /**
+   * @desc The type of connection a device is using to communicate with the network.
+   * It will be one of the following values:
+   *  - bluetooth
+   *  - cellular
+   *  - ethernet
+   *  - none
+   *  - wifi
+   *  - wimax
+   *  - other
+   *  - unknown
+   */
+  type: INetworkInformation['type'] | undefined;
+}
+
+const navigator:
+  | (Navigator &
+      Partial<Record<'connection' | 'mozConnection' | 'webkitConnection', INetworkInformation>>)
+  | undefined = isBrowser ? window.navigator : undefined;
+
+const conn: INetworkInformation | undefined =
+  navigator && (navigator.connection || navigator.mozConnection || navigator.webkitConnection);
+
+function getConnectionState(previousState?: IUseNetworkState): IUseNetworkState {
+  const online = navigator?.onLine;
+  const previousOnline = previousState?.online;
+
+  return {
+    online,
+    previous: previousOnline,
+    since: online !== previousOnline ? new Date() : previousState?.since,
+    downlink: conn?.downlink,
```

Comment:
Why not expose all of `conn` here?

Reply:
`navigator.connection` is a mutable object, also it is EventEmitter instance and it becomes kind of abstraction leak if to simply return it.

Reply:
Got it